### ISSUE TITLE
Rule proposal: `no-return-await`

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#7537]
**Please describe what the rule should do:**
Inside an `async function`, `return await` is useless. Since the return value of an `async function` is always wrapped in `Promise.resolve`, `return await` doesn't actually do anything except add an extra tick before the overarching promise resolves/rejects. This pattern is almost certainly due to programmer ignorance of the return semantics of `async function`s.

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[ ] Warns about a potential error
[x] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
async function foo() {
  return await bar;
}
```

in favor of:
```js
async function foo() {
  return bar;
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**
People are very anxious to use `await`, and will likely overuse it. This rule prevents a common hazard, and also helps educate developers about how `async function` works.

**Comment by not-an-aardvark** [comment]
Seems reasonable to me. :+1:

**Comment by ilyavolodin** [comment]
@ljharb Could you clarify what might be the potential error this rule is going to flag. As in, in which situation returning `await` might cause different behavior than user intended? From what little I know about async/await, I think that it shouldn't cause an error, but `await` is just not necessary for those situations.

**Comment by not-an-aardvark** [comment]
The effect is that the extra `await` will delay the Promise from resolving by an additional tick of the event loop. `await foo;` always executes asynchronously, so placing an `await` in a return statement unnecessarily causes the returned Promise to get fulfilled slightly later.

**Comment by ljharb** [comment]
Exactly as @not-an-aardvark has said (and outlined in the OP).

It won't cause an error - but it will unnecessarily slow down firing of resolution handlers of the resulting promise.

**Comment by nzakas** [comment]
Marking as accepted, as there is a champion and three :+1:s. 

@ljharb can you submit a PR for this?

**Comment by not-an-aardvark** [comment]
I think there are only two :+1:s from team members (excluding mine since I'm championing the issue). I removed my :+1: to avoid confusion.

**Comment by vitorbal** [comment]
I will :+1: this.

**Comment by ilyavolodin** [comment]
In that case, I think this should be "Suggests an alternate way of doing something" rather than "Warns about a potential error". As in, it should go into best "practices section" instead of "possible errors".

**Comment by ljharb** [comment]
@ilyavolodin sure, that sounds great - updated the OP.

@nzakas Happy to submit a PR - so far I've only modified rules, not created them, so this will be a fun one :-)

**Comment by ljharb** [comment]
PR up at https://github.com/eslint/eslint/pull/7547

**Comment by silverwind** [comment]
The [docs](https://github.com/eslint/eslint/blob/master/docs/rules/no-return-await.md) of `no-return-await` link to [async function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) on MDN, which currently features two examples that use `return await`.

This one doesn't trigger the rule. Is that intended?
```js
async function add1(x) {
  var a = resolveAfter2Seconds(20);
  var b = resolveAfter2Seconds(30);
  return x + await a + await b;
}
```

And this one rightfully triggers the rule, I guess it should be fixed on MDN:
````js
async function getProcessedData(url) {
  let v:
  try {
    v = await downloadData(url); 
  } catch (e) {
    v = await downloadFallbackData(url);
  }
  return await processDataInWorker(v);
}
````
Overall, I feel the MDN page isn't detailed enough on how async-await works. Maybe a link to the spec would be better?

**Comment by not-an-aardvark** [comment]
I agree that the MDN page could use some more detail, but I'm not sure linking to the spec instead is a good idea; the spec text can be very difficult to understand for beginners, IMO. (We could link to both, I suppose.)

---

Both of those examples are working as intended. The rule is supposed to flag cases where a value with `await` is *directly* returned from a function; in these cases, the `await` keyword is redundant.

```js
async function foo() {
  // The expression `await resolveAfter2Seconds(5)` has
  // a value of 5. When returned from an async function, the value of 5 is wrapped in
  // a Promise, so the return value is Promise { 5 }.
  return await resolveAfter2Seconds(5);
}

async function foo() {
  // The expression `resolveAfter2Seconds(5)` has
  // a value of Promise { 5 }. This value is returned from the async function, so
  // the return value is Promise { 5 }.
  return await resolveAfter2Seconds(5);
}
```

Since both functions above result in the same return value, `await` is redundant in the first function. However, in your first example this is not the case:

```js
// (slightly simplified from your example)
async function add() {
  var a = resolveAfter2Seconds(20);
  var b = resolveAfter2Seconds(30);
  // `await a` has a value of 20, and `await b` has a value of 30.
  // These are added together resulting in 50, which is wrapped in
  // a Promise and returned from the function. Therefore, the return
  // value is Promise { 50 }.
  return await a + await b;
}

async function add() {
  var a = resolveAfter2Seconds(20);
  var b = resolveAfter2Seconds(30);
  // `a` has a value of Promise { 20 }, and `await b` has a value of Promise { 30 }.
  // These are added together resulting in the string "[object Promise][object Promise]",
  // which is wrapped in a Promise and returned from the function. Therefore, the return
  // value is Promise { "[object Promise][object Promise]" }.
  return a + b;
}
```

`await` is not directly used on the return value in the above example, and it's not redundant (removing it results in a different returned value).

**Comment by silverwind** [comment]
Thanks, now it's clear to me. I [updated](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function$compare?locale=en-US&to=1142087&from=1138589) MDN to remove the `return await` on the last example.

**Comment by syrnick** [comment]
I don't see how it follows from https://tc39.github.io/ecmascript-asyncawait/#abstract-ops-async-function-start. Can you point to the spec defining this behavior?

It seems highly unexpected that an async function implicitly awaits for the promise return value. I wonder if it's in the spec or that's an implementation detail (bug).

E.g. consider

```
async function startExpensiveOperation(opArgs) {
     await startAndWaitUntilOpCompletionAsync(opArgs);
}
async function scheduleExpensiveOperationAsync() {
    const opArgs = await quicklyGetArgsAsync();
    return startExpensiveOperationAsync();
}
```
 
I would expect calling `await scheduleExpensiveOperationAsync()` to be quick, because it doesn't seem to wait for the actual completion.

**Comment by ljharb** [comment]
@syrnick the `return` value of an `async function` absolutely gets wrapped in `Promise.resolve`, which is the same thing as `return await` of that value. In your `startExpensiveOperation`, the resulting promise won't get resolved or rejected until the promise returned from `startAndWaitUntilOpCompletionAsync` settles.

**Comment by not-an-aardvark** [comment]
@syrnick It's implied by 3d. and 3e. of the [AsyncFunctionStart](https://tc39.github.io/ecma262/#sec-async-functions-abstract-operations-async-function-start) (which you linked above).

3a. Let *result* be the result of evaluating asyncFunctionBody.
...
3d. If *result*.[[type]] is normal, then Perform ! Call(promiseCapability.[[Resolve]], undefined, «undefined»).
3e. Else if *result*.[[type]] is return, then Perform ! Call(promiseCapability.[[Resolve]], undefined, «result.[[value]]»)

Translated into English: If the function doesn't return a value, return Promise.resolve(undefined), otherwise return Promise.resolve(theValue).

**Comment by syrnick** [comment]
I would expect `theValue` to be of type promise. 

My understanding is that promiseCapability is the promise constructed at the very beginning of the call to async function.

I have verified that babel with regenerator works according to this eslint rule. I haven't tried any native implementation yet - gotta build it.

**Comment by syrnick** [comment]
Ah, this is not about async functions as much as it is about promises:

https://github.com/tc39/ecmascript-asyncawait/issues/41

"It is impossible currently for a promise to resolve to a promise"

E.g. it will timeout before calling then.

```
var a = new Promise(function(res, rej) {
    var b = new Promise(function(res2, rej2) {
        setTimeout(function() {
            console.log('TIMEOUT');
            res2();
        }, 1000);
    });
    res(b);
}).then(function() {
    console.log('THEN');
});
```

Now this rule makes sense to me. Perhaps you can add the pointer to that issue.

**Comment by emmettnicholas** [comment]
This sentence in the rule description is possibly inaccurate:

> This pattern is almost certainly due to programmer ignorance of the return semantics of async functions.

Another reason someone may prefer `return await` (besides ignorance) is if they were to optimize for readability and consistent code style over perf.

Compare these examples:

```javascript
async fooAsync() {
	const a = await aAsync();
	const b = await bAsync(a);
	return await cAsync(b);
}

// vs

async fooAsync() {
	const a = await aAsync();
	const b = await bAsync(a);
	return cAsync(b);
}
```

The first is arguably slightly more readable, since all 3 lines in the function are similarly formatted (i.e. all use `await`).

Then, if I were to remove `cAsync`'s dependency on `a` and `b`, should I then remove the `async` keyword or leave it in place?

```javascript
async fooAsync() {
	return cAsync();
}

// vs

fooAsync() {
	return cAsync();
}
```

I can imagine someone avoiding this eslint rule to avoid having to make this sort of decision.

**Comment by ljharb** [comment]
@emmettnicholas removing the `async` keyword from a function is a potential breaking change, so effectively there's never a reason to remove it. You'd absolutely want `async fooAsync() { return cAsync(); }`, to give you the strong guarantees that `async function` provides.

**Comment by emmettnicholas** [comment]
> removing the async keyword from a function is a potential breaking change, so effectively there's never a reason to remove it.

I see what you mean, and I'm still trying to formulate my opinion on this, but that guidance runs counter to this new rule: http://eslint.org/docs/rules/require-await. I realize all rules need to go hand in hand, but removing unnecessary keywords is clearly an alternate objective for some.

**Comment by ljharb** [comment]
@emmettnicholas yes, `require-await` imo is a harmful and bad rule, and there's never a good reason to enable it. I argued against its inclusion but the TSC decided to include it ¯\\\_(ツ)_/¯

**Comment by gkatsanos** [comment]
I run into this use case: 

```js
async passwordMatches(password) {
  return await bcrypt.compare(password, this.password); // eslint-disable-line no-return-await
}

  async findAndGenerateToken(options) {
    const { email, password, refreshObject } = options;
    const user = await this.findOne({ email }).exec();
    if (password) {
      if (user && await user.passwordMatches(password)) {
        return { user, accessToken: user.token() };
      }
      err.message = 'Incorrect email or password';
    }
```
how should that be written in order to be ok with `no-return-await` AND `no-useless-return` ? 
note: brypt.compare [returns a Promise](https://github.com/dcodeIO/bcrypt.js#compares-hash-callback-progresscallback)

**Comment by ljharb** [comment]
@gkatsanos just `return bcrypt.compare(…)`

**Comment by gkatsanos** [comment]
@ljharb check updated Q :) thanks

**Comment by ljharb** [comment]
Similarly, `return window.fetch(…)`.

All returns in an async function are wrapped in `Promise.resolve`, so `return await` is redundant.

You can't have a sync function that does anything based on anything async - every single thing that depends on an async thing needs to also be async, either by taking a callback or returning a promise (as `async function` does).

**Comment by gkatsanos** [comment]
so specifically in `findAndGenerateToken()`, it will work since it's declared as `async`. But in ```if (user && await user.passwordMatches(password)) {``` is `await` necessary? (my code wasn't working before I add it, the condition was evaluated as true)

**Comment by ljharb** [comment]
You definitely need the `await` there, yes.

**Comment by shinout** [comment]
I found a case we need `return await`.

```js

async function foo() {
  try {
    return await bar()
  }
  catch (e) {
    console.log(e) // error comes here
  }
}

async function fooWithoutAwait() {
  try {
    return bar()
  }
  catch (e) {
    console.log(e) // cannot catch async error
  }
}

async function bar() {
  throw new Error('bar')
}

foo() // catch error
fooWithoutAwait() // unhandled promise rejection
```

How can we manage these async exceptions?

**Comment by not-an-aardvark** [comment]
@shinout The `no-return-await` rule does not report those cases, so you should be able to use them as you've written them there.

**Comment by shinout** [comment]
Thanks. Sorry for not checking the rule behavior.
